### DISCUSSION TITLE
Option to Filter Visible Programming Blocks (for Younger Students in a Classroom)

### DISCUSSION THREAD

**Discussion by Ben-Toolbox** [pybricks/support#1689] [Ideas]
I think pybricks has the potential to be an amazing teaching tool for students of all ages.  I am still exploring how it may work with my younger students - right now I am holding back exposing them to it, as I know they would get overwhelmed by all of the options and functionality.

As a teaching tool, I would love to customize my own chosen menu of pybricks programming blocks for a given class, maybe even sprinkling in some custom blocks that I create as options for the kids.

I understand if shooting for younger audiences is not a main priority for pybricks, but I think the depth of options already there gives teachers a HUGE resource to choose from.  How amazing would it be to create a curriculum built with limited menus of blocks to slowly expose students to the depth of pybricks!  You could potentially guide kids as young as 5 all the way through high school and beyond.

For years, pybricks has offered a much clearer, deeper alternative to python programming when compared to Spike Prime python interface.  Now, the new block programming has done the same when compared to block based coding in Spike.  I am now interested in the possibility of being able to customize a smarter, more versatile alternative to the Spike Icon Block programming option, which I feel is completely lacking.  

Would love to hear your thoughts, and thanks for your time!

**Comment by laurensvalk** [comment]
> I think pybricks has the potential to be an amazing teaching tool for students of all ages. 

Thank you!

> I understand if shooting for younger audiences is not a main priority for pybricks, but I think the depth of options already there gives teachers a HUGE resource to choose from. How amazing would it be to create a curriculum built with limited menus of blocks to slowly expose students to the depth of pybricks! You could potentially guide kids as young as 5 all the way through high school and beyond.

This is certainly something I'd like to do. We're a very small team though, so we try to focus on one thing at a time. This is why the block coding option came relatively late.

> I am now interested in the possibility of being able to customize a smarter, more versatile alternative to the Spike Icon Block programming option (...)  I would love to customize my own chosen menu of pybricks programming blocks for a given class, maybe even sprinkling in some custom blocks that I create as options for the kids

Thanks for sharing this. This is something I've been thinking about a long time, particularly after playing with the SPIKE Essential set with my young kids. It would be nice if they could simply connect a "when helicopter moves" block and a "spin the blades" block, instead of learning about hub tilt or motors right away. And when they're curious, that they could see which blocks are inside.

The way Pybricks is set up, it lends itself quite well to such abstractions. The DriveBase is one example, but still a bit too advanced for complete beginners. I think we could potentially take that a step further and have dedicated, lesson-specific blocks like the helicopter example above.

It may be a while before we get there, though.

**Reply by Ben-Toolbox** [reply]
Yes, what you describe with the 'when helicopter moves' block is just what I would imagine.... allowing the kids to use a bundle of code to understand the basics and then unpack it when they are ready.

Or even simpler than having it subject-specific (like the Drive Base or the helicopter), could it be possible to provide tools so that teachers could recreate a custom block like a "Start on Button Pressed" hat block or a "Move Motor for ___ Seconds" block.  Teachers could essentially recreate, customize and improve upon these simple blocks that are in some of the Spike software, and use them for younger kids.

Much of this can already be approached with making custom tasks.  Again, for my younger kids, so much of this is about NOT seeing all the options so they can manipulate and understand only a few elements at a time.  This is where I would imagine (eventually, of course!!) some kind of custom checklist / dropdown feature to select which blocks or sections are available in any given class 'palette'... so if we are not dealing with 'Drive Base, Car, Communication'  those sections could be hidden or collapsed in some way.

An example of the limitations of Spike Essential / Prime Icon Block coding is how there is no easy way to just turn a motor on continuously.  So if the kids just want to turn the motor on with a green light, then wait for red to turn it off, they can't easily do it (unless, of course I am missing something!).  This makes the whole platform essentially unusable for me in a class of younger kids.

> I think we could potentially take that a step further and have dedicated, lesson-specific blocks like the helicopter example above.

The big picture is that I feel that the above task could potentially be done by teachers using your tools, rather than you all at pybricks needing to make a bunch of lesson-specific blocks.  

Thanks for your thoughtful response, Laurens.  Keep up the amazing work you all continue to do!

-Ben
